### ISSUE TITLE
Proposal: allow webRequest with `background.persistent: false` / SW

### ISSUE THREAD

**Issue by carlosjeurissen** [w3c/webextensions#148]
Currently the webRequest and webRequestBlocking permission require a persistent background page. This seems to be an arbitrary limitation. The proposal is to allow these APIs under a non-persistent / SW API as well.

**Comment by Rob--W** [comment]
Could you clarify, why do you believe that this API requires a persistent background page? The use cases for extension popups/tabs sound reasonable, and I'd imagine the API to work in extensions without a background script (persistent or not).

I know that there were some technical reasons (implementation details) that prevented (blocking) webRequest from being supported in Chrome when event pages are enabled, but does this issue also extend to other browsers? Is this issue a bug report about Chrome, or is it broader than that?

**Comment by carlosjeurissen** [comment]
@Rob--W Currently I'm not aware of any reasoning behind this. I can only speculate on that. Potentially it was seen that webRequest listeners would mostly make sense when they are constantly there. But that is not necessarily true.

**Comment by Rob--W** [comment]
> ..., but does this issue also extend to other browsers? Is this issue a bug report about Chrome, or is it broader than that?

@carlosjeurissen ^

**Comment by carlosjeurissen** [comment]
@Rob--W Just tested in Safari on macOS and turns out just like in Google Chrome one cannot use the webRequest API with a non-persistent background page.

As for Mozilla Firefox, not sure what the current behaviour is with the WebRequest API and the limited event page (https://github.com/w3c/webextensions/issues/134)

**Comment by Rob--W** [comment]
This topic came up during today's WECG meeting. Relevant part from the meeting notes (currently pending review, will be merged and published at https://github.com/w3c/webextensions/blob/main/_minutes/2022-02-03-wecg.md)

> * [timothy] webRequest is not supported in non-persistent background pages in Safari, because we looked at Chrome, which didn't support it either.

**Comment by tophf** [comment]
Chrome didn't support webRequest in an event page because it wasn't exactly trivial to support waking up the event page and it would cause frequent restarts of the event page as such events tend to happen frequently.

The support for SW, which are conceptually the same as event pages, is being implemented currently in https://crbug.com/1024211, so there's no reason it can't be made to work with an event page. Note that webRequest + webRequestBlocking as *optional permissions* always worked in an event page while it was running, it just didn't wake up.

**Comment by carlosjeurissen** [comment]
@tophf Thanks for clarifying! My use case for the webRequest API is only in non-background contexts. In which case it's a pity a permanent background page has to run for this to work.

**Comment by tophf** [comment]
But you don't need the background script at all if you use webRequest only in a visible page.

**Comment by carlosjeurissen** [comment]
@tophf correct! Yet Google Chrome doesn't allow loading an extension with the webRequest permission and a non-persistent background page at this moment.

**Comment by tophf** [comment]
As I mentioned in my comment, all you need is to use [optional permissions](https://developer.chrome.com/extensions/permissions) and it'll work with a non-persistent background script.

**Comment by carlosjeurissen** [comment]
Firefox currently allows loading extensions with the`webRequest` permission when `persistent` is set to `false` in mv2. In mv3, the `persistent` key is unsupported.

Related: a proposal for adding support for the webRequest API in non-persistent background pages has been added to the chrome issue tracker by @Rob--W https://bugs.chromium.org/p/chromium/issues/detail?id=447270

**Comment by dotproto** [comment]
2026 London F2F triage

@zombie confirmed that WebRequest is supported in Firefox with event pages.
@carlosjeurissen indicated that Safari doesn't support webRequest event listeners with `persistent: false`.

**Comment by rdcronin** [comment]
This is implemented in Chrome (for service worker contexts).  The blocking capabilities of webRequest are limited to policy-installed extensions, but that's orthogonal to this issue.

**Comment by carlosjeurissen** [comment]
Just verified in Safari. And this is now supported as well in Safari. Closing the issue as it is implemented in all browsers now!
